libraries/transport/transportControl/transportControlI.H

```diff
@@ -158,31 +148,51 @@ void Foam::Pmt::transportControl<Base>::update()
                 [](scalar v){ return v>=0; }
             );
 
+        dtState_ = dtState::SHOULD_INCREASE;
 
         forAll(composition_.Y(), speciesi)
         {
             const auto& Y = composition_.Y(speciesi);
+            const auto& Y_old = composition_.Y(speciesi).oldTime();
 
             if (maxDeltaCSet && relMaxDeltaCSet)
             {
-                maxDeltaY_[speciesi] = max(relMaxDeltaC*gMax(Y), maxDeltaC[speciesi]);
+                maxDeltaY[speciesi] = max(relMaxDeltaC*gMax(Y), maxDeltaC[speciesi]);
             }
             else if (maxDeltaCSet)
             {
-                maxDeltaY_[speciesi] = maxDeltaC[speciesi];
+                maxDeltaY[speciesi] = maxDeltaC[speciesi];
             }
             else if (relMaxDeltaCSet)
             {
-                maxDeltaY_[speciesi] = max(relMaxDeltaC*gMax(Y), SMALL);
+                maxDeltaY[speciesi] = max(relMaxDeltaC*gMax(Y), SMALL);
             }
             else
             {
-                maxDeltaY_[speciesi] = std::numeric_limits<scalar>::infinity();
+                maxDeltaY[speciesi] = std::numeric_limits<scalar>::infinity();
             }
 
-            maxDYdT_[speciesi] = gMax(mag(composition_.ddtScheme(speciesi)->fvcDdt(Y))());
-        }
+            //use the iterations to calculate field variations and decide state
+
+            auto residual = gMax((mag(Y-Y_old))->internalField());
 
-        upToDate_ = true;
+            Info << "Species " << composition_.species()[speciesi] << " variation: "<< residual << endl;
+
+            if (maxDeltaY[speciesi] != std::numeric_limits<scalar>::infinity()){ //because order of SPECIES's maxDeltaC could affect
+
+                if (dtState_ != dtState::SHOULD_DECREASE)   //Implementing a finite state machine
+                {
+                    if (residual >= 1.*maxDeltaY[speciesi]){
+                        dtState_ = dtState::SHOULD_DECREASE;
+                    } 
+                    else if (residual >= 0.75*maxDeltaY[speciesi]){
+                        dtState_ = dtState::SHOULD_MAINTAIN;
```

Comment:
Are we sure this couldn't erroneously change a `SHOULD_DECREASE` set by one species into a `SHOULD_MAINTAIN`?